### PR TITLE
Work with model architectures

### PR DESCRIPTION
**Pull request by matekrk**
1. Split massive model.py into dir models
2. Add documentations for each method and class definition
3. Last fixes with CC (especially logistic VI)
4. Correct support for VBLL (for diag/lowrank/dense)
5. Remove activation in the end of backbone?

### CODE REVIEW

File: src/models/backbone.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    """
+
+    p = cfg.p if hasattr(cfg, 'p') else 64
+    data_channels = cfg.data_channels if hasattr(cfg, 'data_channels') else 3
+    if cfg.backbone_type == 'ConvNet':
+        backbone = ConvNet(data_channels, p).to(torch.double)
+    elif cfg.backbone_type == 'MLP':
+        backbone = MLP(data_channels, cfg.data_size, p).to(torch.double)
+    elif cfg.backbone_type == 'ResNet34':
+        import torchvision.models as models
```
Comment (tkusmierczyk) [inline src/models/backbone.py]: no name conflicts `torchvision.models` vs `models`?

File: src/models/logistic_point.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+            assert isinstance(prior_mu, torch.Tensor), "mu must be a torch.Tensor"
+            assert prior_mu.shape[0] == self.K, f"mu must have shape ({self.K}, p)"
+            # assert prior_mu.shape[1] == self.p, f"mu must have shape ({self.K}, {self.p})" turn off for CC
+            self.prior_mu_list = [prior_mu[k] for k in range(self.K)]
+
+        if prior_Sig is None:
+            self.prior_Sig_list = [torch.eye(self.p, dtype=torch.double) for _ in range(self.K)]
+        else:
+            assert isinstance(prior_Sig, torch.Tensor) or isinstance(prior_Sig, list), "Sig must be a list of tensors"
+            if isinstance(prior_Sig, torch.Tensor):
```
Comment (tkusmierczyk) [inline src/models/logistic_point.py]: Do we ever use it? Would it even work if we passed a full 3D tensor here?

File: src/models/logistic_point.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+import torch.nn as nn
+from .generic import LLModel, LLModelCC
+
+"""## Sigmoid-pointwise model"""
+class LogisticPointwise(LLModel):
+    """
+    Logistic regression model for pointwise predictions. This model is used for binary classification tasks where each output is independent.
+    """
+    def __init__(self, p, K, beta=1.0, intercept=False, backbone=None, m_init=None, 
+                 prior_mu=None, prior_Sig=None, prior_mean_learnable=False, prior_scale_learnable=False):
```
Comment (tkusmierczyk) [inline src/models/logistic_point.py]: `prior_mean_learnable` and `prior_scale_learnable` does not seem to be saved anywhere!?!

File: src/models/logistic_point.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+                    assert m.shape == (self.p,), f"m_init[{i}] must have shape ({self.p},)"
+            self.m_list = [nn.Parameter(m_init[:, k]) for k in range(self.K)]
+
+    
+    def get_learnable_parameters(self):
+        params = nn.ParameterList(self.m_list)
+        if self.backbone is not None:
+            params += list(self.backbone.parameters())
+        if self.prior_mean_learnable:
+            self.prior_mu_list = [nn.Parameter(mu) for mu in self.prior_mu_list]
```
Comment (tkusmierczyk) [inline src/models/logistic_point.py]: How does this suppose to work? At what point we make parameters of the prior learnable?

File: src/models/logistic_point.py [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+        Notes:
+            - The `data_size` parameter is used to scale the regularization term. If not provided, the batch size is used.
+            - The method computes the BCE loss for the predictions and adds the regularization term scaled by the `beta` parameter or `other_beta` if provided.
+            - If `verbose` is True, the method prints the BCE loss and the regularization term (if `beta` or `other_beta` is non-zero).
+        """
+        data_size = data_size or X_batch.shape[0]
+
+        preds = self.forward(X_batch)
+        assert preds.shape == y_batch.shape, f"preds.shape={preds.shape} != y_batch.shape={y_batch.shape}"
+        mean_bce = self.loss(preds, y_batch)
```
Comment (tkusmierczyk) [inline src/models/logistic_point.py]: `self.loss` has set reduction to `mean`. Does this take average both over data and the K labels? 
If so, it is K times smaller than log-likelihood which is bugs prone.

**Comment by matekrk** [comment]
In the future PRs:
- improve logic behind order of heads depending on the order of CC for pointwise / VI models
- make main.py file slimmer & put similarly to data dir relevant python files
- give configs to correct runs

File: src/models/vi.py [hunk 6]
```diff
@@ -0,0 +1,856 @@
...
+            - The `other_beta` parameter allows overriding the default regularization parameter `self.beta` for specific computations.
+        """
+        X_processed = self.process(X_batch)
+        batch_size = X_batch.shape[0]
+
+        m_list = [m.to(X_batch.device) for m in self.m_list]
+        prior_mu_list = [mu.to(X_batch.device) for mu in self.prior_mu_list]
+        y_list = [y_batch[:, val_k] for i_k, val_k in enumerate(range(self.K))]
+
+        if self.method in [0, 4]:
```
Comment (tkusmierczyk) [inline src/models/vi.py]: add comment saying what is 0 and what is 4

File: src/models/vi.py [hunk 7]
```diff
@@ -0,0 +1,856 @@
...
+            prior_Sig_list = [sig.to(X_batch.device) for sig in self.prior_Sig_list]
+
+            if self.method == 0:
+                likelihood = -neg_ELL_TB_MH(m_list, s_list, y_list, X_processed, l_max=self.l_terms)
+                KL_div = KL_MH(m_list, s_list, prior_mu_list, prior_Sig_list)
+            else:
+                likelihood = -neg_ELL_MC_MH(m_list, s_list, y_list, X_processed.to(X_batch.device), n_samples=self.n_samples)
+                KL_div = KL_MH(m_list, s_list, prior_mu_list, prior_Sig_list)
+
+        elif self.method in [1, 5]:
```
Comment (tkusmierczyk) [inline src/models/vi.py]: add comment saying what is 1 and what is 5

File: src/models/vi.py [hunk 8]
```diff
@@ -0,0 +1,856 @@
...
+        m_list = [m.to(X_batch.device) for m in self.m_list]
+        y_list = [y[:, i_k] for i_k in range(self.K)]
+        if self.method in [0, 4]:
+            s_list = [s.to(X_batch.device) for s in self.s_list]
+            if mc:
+                nlls = [neg_ELL_MC(m, s, y, X_processed, n_samples=n_samples) for m, s, y in zip(m_list, s_list, y_list)]
+            else:
+                nlls = [neg_ELL_TB(m, s, y, X_processed, l_max=self.l_terms) for m, s, y in zip(m_list, s_list, y_list)]
+        elif self.method in [1, 5]:
+            S_list = [S.to(X_batch.device) for S in self.S_list]
```
Comment (tkusmierczyk) [inline src/models/vi.py]: Why do we keep changing the devices?

File: src/models/vi.py [hunk 9]
```diff
@@ -0,0 +1,856 @@
...
+        Notes:
+            - The NLL is computed for each output (attribute) separately.
+            - If `mc` is True, Monte Carlo estimation is used to compute the NLL.
+            - The method supports both diagonal and full covariance matrices, depending on the model configuration.
+        """
+        X_processed = self.process(X_batch)
+        m_list = [m.to(X_batch.device) for m in self.m_list]
+        y_list = [y[:, i_k] for i_k in range(self.K)]
+        if self.method in [0, 4]:
+            s_list = [s.to(X_batch.device) for s in self.s_list]
```
Comment (tkusmierczyk) [inline src/models/vi.py]: Why do we keep changing the devices?

File: src/models/vi.py [hunk 10]
```diff
@@ -0,0 +1,856 @@
...
+
+        Notes:
+            - For diagonal covariance matrices, the method uses either a probit approximation or Monte Carlo sampling.
+            - For full covariance matrices, the method uses either a probit approximation or Monte Carlo sampling.
+            - The probit approximation is based on the paper "Probit Approximation for Bayesian Logistic Regression"
+              (https://arxiv.org/pdf/2002.10118).
+            - Monte Carlo estimation is used when `mc` is True, and the number of samples is specified by `n_samples`.
+        """
+        assert (not mc or n_samples is not None), "n_samples must be provided for Monte Carlo estimation"
+        m = self.m_list[i_relevant].to(X_batch_processed.device)
```
Comment (tkusmierczyk) [inline src/models/vi.py]: why do we keep flipping the devices??

File: src/models/vi.py [hunk 11]
```diff
@@ -0,0 +1,856 @@
...
+        
+        M = X_batch_processed @ m # just take mean of the distribution
+        if self.incorrect_straight_sigmoid:
+            return torch.sigmoid(M), M
+
+        if self.method in [0, 4]:
+            u = self.u_list[i_relevant].to(X_batch_processed.device)
+            s = torch.exp(u)
```
Comment (tkusmierczyk) [inline src/models/vi.py]: isn't it equivalent to self.s_list[i_relevant]?

File: src/models/vi.py [hunk 12]
```diff
@@ -0,0 +1,856 @@
...
+            if not mc: # probit approximation
+                scaling_factor_diag = torch.einsum("bi,i,bi->b", X_batch_processed, s**2, X_batch_processed)
+                assert scaling_factor_diag.shape == torch.Size([X_batch_processed.shape[0]])
+                scaling_factor = torch.sqrt(1 + (torch.pi / 8) * scaling_factor_diag)
+                M_corrected = M / scaling_factor
+                expected_sigmoid = torch.sigmoid(M_corrected)
+            else: # ref objective.py 142
+                S = torch.sqrt(torch.sum(X_batch_processed**2 * s**2, dim=1))
```
Comment (tkusmierczyk) [inline src/models/vi.py]: torch.sqrt of torch.sqrt? WTF?

File: src/models/vi.py [hunk 13]
```diff
@@ -0,0 +1,856 @@
...
+        self.prior_scale_learnable = prior_scale_learnable
+
+        self.incorrect_straight_sigmoid = incorrect_straight_sigmoid
+        self.sigmoid_mc_computation = sigmoid_mc_computation
+        self.sigmoid_mc_n_samples = sigmoid_mc_n_samples
+
+        # Initialize prior parameters
+        if prior_mu is None:
+            self.prior_mu = nn.Parameter(torch.tensor(0.0, dtype=torch.double), requires_grad=self.prior_mean_learnable)
+        else:
```
Comment (tkusmierczyk) [inline src/models/vi.py]: do we ever use it? do we need it?

File: src/models/vi.py [hunk 14]
```diff
@@ -0,0 +1,856 @@
...
+    @property
+    def prior_Sig_list(self):
+        """
+        Return the prior covariance matrices for each output.
+        """
+        ps = self.prior_scale
+        return [torch.eye(self.p, dtype=torch.double, device=ps.device) * ps for _ in range(self.K)]
+
+    @property
+    def s_list(self):
```
Comment (tkusmierczyk) [inline src/models/vi.py]: we define it and it does not seem we use it

File: src/models/vi.py [hunk 15]
```diff
@@ -0,0 +1,856 @@
...
+        likelihood = torch.tensor(0.0, dtype=torch.double, device=X_batch.device)
+        KL_div = torch.tensor(0.0, dtype=torch.double, device=X_batch.device)
+
+        prev_list = []
+        for i_k, val_k in enumerate(self.chain_order):
+            i_relevant = (self.chain_order == i_k).nonzero().item()
+            if i_k == 0:
+                X = X_processed
+            else:
+                X = torch.cat((X_processed, torch.cat(prev_list, dim=1)), dim=1)
```
Comment (tkusmierczyk) [inline src/models/vi.py]: why do we perform concatenation of everything in every step?

**Review by tkusmierczyk — CHANGES_REQUESTED** [review]
remove the .to() statements!

**Comment by matekrk** [comment]
this is updated (especially VI) code
